### DISCUSSION TITLE
AI: Abstraction

### DISCUSSION THREAD

**Discussion by mhasanunal** [photoprism/photoprism#4577] [Ideas]
It may be a good idea to make AI handling via abstract way?
This way, we can call a microservice or current implementation?

There should be a DI, an interface where implements:
input is a image in binary, 
output is embeddings and/or labels/rectange/point.

and even we can change the strategy on-the-fly?

**Comment by graciousgrey** [comment]
We've already started working on a micro-service that will allow us to use advanced computer vision models through a REST API:

https://github.com/photoprism/photoprism-vision

**Comment by mhasanunal** [comment]
That's great! But consider abtraction too. Because let say I already have a face recognition backend with various options(models).

I can always write a proxy API to match photoprism interface...

**Comment by hkorte** [comment]
Hi, I’m also looking for a way to use embeddings for search to enable queries like “Two children playing with a dog in front of a church in Ecuador.” Is this also the idea behind "PhotoPrism Vision"? Is it already usable? How can this be activated in PhotoPrism?
